src/features/device-state/routes/state.ts

```diff
@@ -25,6 +26,161 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
 	});
 };
 
+export type App = {
+	name: string;
+	commit: string;
+	releaseId: number;
+	appId?: string;
+	uuid?: string;
+	services: {
+		[id: number]: AppService;
+	};
+	volumes: AnyObject;
+	networks: AnyObject;
+};
+
+export type AppService = {
+	imageId: number;
+	serviceName: string;
+	image: string;
+	running: boolean;
+	environment: Dictionary<string>;
+	labels: Dictionary<string>;
+	contract?: string;
+};
+
+export type LocalState = {
+	name: string;
+	config: Dictionary<string>;
+	apps: Dictionary<Partial<App>>;
+};
+
+async function buildAppFromRelease(
+	device: AnyObject,
+	application: AnyObject,
+	release: AnyObject,
+	config: Dictionary<string>,
+): Promise<App> {
+	let composition: AnyObject = {};
+	const services: App['services'] = {};
+
+	// Parse the composition to forward values to the device
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
```

Comment:
fwiw I don't think this `JSON.parse` should ever be useful - I'd expect it to either already be parsed or potentially have had a null value in the db (although I don't think it's still non-nullable?). I'd leave it for a separate PR though along with verification that it's definitely not necessary

Reply:
This is not a new change.  That code was there like that from before.   It only popped up now in the refactor because it was moved around.

Reply:
> although I don't think it's still non-nullable

@Page- it is still non-nullable, FYI -- I feel that `{}` is as good to represent "no composition" and less error-prone (ie. code always reads valid JSON).